File: frontend/scripts/check-i18n.mjs
import { readFileSync, readdirSync, statSync } from "node:fs";
import { join } from "node:path";

const roots = ["app", "components", "features", "lib"].map((path) => join("frontend", path));
const allowed = new Set([join("frontend", "lib", "i18n.ts")]);
const localizedText = /[\u0400-\u04ff\u10a0-\u10ff]/;
const checkedExtensions = new Set([".ts", ".tsx"]);

const violations = [];

for (const root of roots) {
  scan(root);
}

if (violations.length > 0) {
  console.error("Hardcoded localized UI text found outside dictionaries:");
  for (const violation of violations) {
    console.error(`- ${violation}`);
  }
  process.exit(1);
}

function scan(path) {
  let stats;
  try {
    stats = statSync(path);
  } catch {
    return;
  }
  if (stats.isDirectory()) {
    for (const item of readdirSync(path)) {
      scan(join(path, item));
    }
    return;
  }
  if (!checkedExtensions.has(path.slice(path.lastIndexOf(".")))) {
    return;
  }
  if (allowed.has(path)) {
    return;
  }
  const content = readFileSync(path, "utf8");
  if (localizedText.test(content)) {
    violations.push(path);
  }
}

File: app/scripts/check_i18n.py
from __future__ import annotations

import re
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parents[2]
APP_ROOT = ROOT / "app"
LOCALIZED_TEXT = re.compile(r"[\u0400-\u04ff\u10a0-\u10ff]")
ALLOWED = {
    APP_ROOT / "i18n.py",
    APP_ROOT / "reference_data.py",
}
ALLOWED_DIR_PARTS = {"alembic", "tests", "__pycache__"}


def main() -> int:
    violations: list[str] = []
    for path in APP_ROOT.rglob("*.py"):
        if path in ALLOWED or any(part in ALLOWED_DIR_PARTS for part in path.parts):
            continue
        if LOCALIZED_TEXT.search(path.read_text(encoding="utf-8")):
            violations.append(str(path.relative_to(ROOT)))
    if violations:
        print("Hardcoded localized backend text found outside i18n/reference dictionaries:")
        for violation in violations:
            print(f"- {violation}")
        return 1
    return 0


if __name__ == "__main__":
    sys.exit(main())
